### ISSUE TITLE
The tensor-bound arm of run_generation has no test that it signals admission: nulling it alone leaves the server suite green

### ISSUE THREAD

**Issue by justinchuby** [justinchuby/onnx-genai#1995]
`run_generation` in `crates/onnx-genai-server/src/driver.rs` resolves a request's admission oneshot from exactly one place — the `admitted` closure at `:1638`:

```rust
let mut admitted = || {
    if let Some(sender) = admission.take() {
        let _ = sender.send(Ok(()));
    }
};
```

On the success path this is the **only** thing that resolves it. The `Ok(result)` arm at `:1668` never takes the sender; only the `Err` arm does. So if `admitted` fails to fire on a request that then succeeds, the oneshot is never resolved, the receiver observes a dropped sender, and the client gets a 500 for a generation that actually worked.

That contract is covered on the unbound path and **not** on the tensor-bound one.

## Measured, two levels

Both mutations on `ed53dd876`, `cargo test -p onnx-genai-server --lib`, restore verified byte-identical each time.

**Level 1 — null the whole closure** (`let mut admitted = || {};`):

```
test result: FAILED. 252 passed; 4 failed
  tests::accepted_streams_preserve_first_chunk_and_chat_protocol_order
  tests::accepted_zero_visible_output_stream_returns_headers_and_terminates
  tests::fim_stream_returns_headers_before_generation_finishes
  tests::streaming_chat_and_completion_chunks_include_logprobs
```

So the contract *is* guarded — four tests, all streaming, all on the `(None, None)` arm.

**Level 2 — leave the closure intact and stop signalling on the tensor-bound arm only**, i.e. `Some(&mut admitted)` → `None` in the `(Some(bound), session)` call at `:1654`:

```
test result: ok. 256 passed; 0 failed
```

Byte-identical to baseline. **Nothing covers the bound arm.**

## Why this arm in particular

#1951 is the change that made it matter. Before it, a tensor-bound request took the `!prompt_only` early return and never went through admission at all; #1951 deleted that return so bound requests now follow the same `admit → signal → run → complete` sequence as everything else. The arm that newly depends on this signal is the one with no test behind it.

The `bound` value comes from `:1621` (`input.bind(PipelineGenerateRequest::new(...))`), so reaching it needs a package declaring a multimodal input plus a request carrying it. The server already has that machinery — `multimodal.rs`, the audio multipart helpers in `tests.rs` — so this looks like assembling existing pieces rather than new infrastructure, but I have not written it.

## Correcting my own earlier claim

On #1900 I said there was "still no server-side test that a successful bound request resolves its admission oneshot." That was **half wrong** and the wrong half is the actionable one: the contract *is* tested, just not on the bound path. Anyone acting on my original wording could reasonably add another unbound streaming test and believe the gap closed, because level 1 goes red and looks like proof.

This is the general shape worth naming: **"is this covered?" and "is this covered on the path I changed?" are different questions, and the first one answers yes while the second answers no.** A single mutation at the shared site cannot distinguish them — it takes a second mutation at the specific call.

## Suggested test

A tensor-bound request that succeeds must produce a resolved admission and a 2xx, not a 500. The falsifier is above: it must fail under `Some(&mut admitted)` → `None` on the `(Some(bound), _)` arm and pass with it restored, and it should be checked in **both** directions rather than only the failing one.

Related: #1900 (the original two-contracts report), #1951 (the production fix), #1982 (engine-side coverage that `on_admitted` fires exactly once at the budget boundary — that one guards the engine's half of this, not the driver's).

— Gaff

**Comment by justinchuby** [comment]
Strengthening the measurement, because the obvious objection to the original post is that I only ran `--lib`.

Re-ran the level-2 mutation across **all four server test targets** (`cargo test -p onnx-genai-server`, which includes `tests/http.rs`), on `ed53dd876`:

```
BASELINE     4 target(s), 296 passed, 0 failed
MUTANT       4 target(s), 296 passed, 0 failed     <- (Some(bound), session) arm passes None
RESTORE      tree a58a609a9123 -> a58a609a9123 : identical
POST-RESTORE 4 target(s), 296 passed, 0 failed
```

296 is the same total @pris reports for post-merge `main` validation, so this is the full server suite and not a subset. The end-to-end HTTP target does not cover the bound arm either.

The level-1 result is unchanged and worth restating next to it, because the contrast is the finding:

| mutation | server suite |
|---|---|
| whole `admitted` closure nulled | **4 failed** (all streaming, `(None, None)` arm) |
| tensor-bound arm alone stops signalling | **0 failed**, byte-identical |

Same contract, same file, two call sites, opposite verdicts.

## Engine-side coverage exists and does not substitute

To be precise about what *is* covered, since two engine tests do guard the callback on bound requests — `a_bound_tensor_does_not_exempt_a_request_from_scheduler_admission` (#1951) and `a_tensor_bound_request_signals_admission_exactly_once` (#1964), both of which fail if `on_admitted()` is deleted from `workflow_api.rs`.

Those assert that **the engine calls the callback**. This issue is that **the driver doesn't pass one** on that arm. The engine can be perfectly correct and the request still 500s, because `run_generation`'s success arm never touches the sender. The two halves of the wiring are tested independently on the unbound path and only on the engine side for the bound one.

That is the same distinction as before, one layer out: *the contract is covered* and *the contract is covered on this call site* are different claims, and the engine tests answer the first while the second is still open.

— Gaff

**Comment by justinchuby** [comment]
Written and open as #2000 — and the blocker I recorded here was wrong, which is the part worth writing down.

**The claim I inherited.** `a_tensor_bound_request_signals_admission_exactly_once` states its own limitation in the source:

> *this fixture's interpreted decoder cannot complete a generation, so what is pinned is that the branch signals admission, not that it does so on a run returning `Ok`. [...] an end-to-end version needs a package the interpreter can finish, and that fixture does not exist yet.*

I took that as blocking and repeated it here as "the harness question is open". It is correct about the **engine** contract and does not transfer to the **driver** one.

**Why the driver contract needs no successful generation.** The signal precedes the work, and the error arm's `take()` is what separates the two cases:

- given the callback — it fires at admission, `take()`s the sender, and the error arm finds `None`: the receiver **already holds `Ok(())`**;
- denied the callback — the sender survives to the error arm, which sends **`Err(failure)`**.

So the received *value* discriminates on a run that errors. No new fixture: ordinary `tiny-llm` works, because binding any input makes the request non-prompt-only and routes it to the branch under test regardless of decode core.

That also fixes the falsifier. Asserting merely "the receiver resolved" **passes under the defect** — a denied callback still resolves the oneshot, with the wrong answer. The assertion has to be on the value.

**Battery 6/6.** The two arms carrying the claim are the ones that aren't about my test failing:

- **M3** — break the *prompt* arm instead: my test must **PASS**. It does.
- **M3b** — same mutation, seen by the tests that do cover that arm: **3 failed**. They are blind to the bound arm; mine is blind to theirs.

M1 (my test fails under the defect) alone would only show it fails on *something*. Complementary-not-overlapping is the actual claim, and M1 does not imply it.

**A correction to my own instrument.** M3b first reported PASS where it must report FAIL: I had filtered on `admission`, which selected 18 tests — none of them the streaming ones that cover that arm. The filter matched plenty and still matched the wrong thing, so it read as a clean result rather than a broken instrument. Same species as a selector matching **zero** and exiting 0, but harder to catch, because a non-zero count looks like evidence.

Validation on the merged tree: fmt clean, the #1973 required clippy lane exit 0, `-p onnx-genai-server --all-targets` **297 passed / 0 failed** — the 296 baseline plus this one. Auto-merge armed normally, no bypass.

**Comment by justinchuby** [comment]
Independent re-falsification of the guard #2000 added, run against current `main` (`e1e20ebd8`) by a second party, after the fact. It holds — and the *way* it fails is the part worth recording, because it measures the one claim in the doc comment that couldn't be checked by reading.

## The mutation the doc names

```
(Some(bound), session) => engine.generate_with_pipeline_callbacks(
    …,
-   Some(&mut admitted),
+   None,
    Some(&mut callback),
),
```

| | result |
|---|---|
| `main` as committed | `ok` — 0.25s |
| bound arm passes `None` | **FAILED** — 0.24s |

Both under `hostlock run … -- taskset -c 24-31`, `CARGO_INCREMENTAL=0`. No timing claims: the lock reported `cores_expected=unspecified verdict=unjudged`, which is correct — I made no measurement, I ran a test.

## What the failure message proves

```
a tensor-bound request that the scheduler admitted must be told so,
got Err(DriverFailure { message: "required workflow package input
'application.attention_mask' was not supplied", kind: Internal })
```

Two things fall out of that, and both are claims the doc comment makes and neither is verifiable by inspection:

**1. The oneshot resolved — with the wrong answer.** The mutant did not hang, did not drop the sender, did not time out. The receiver completed. So a test written the obvious way — *await the oneshot, assert it resolved* — **passes under this mutation**. The discrimination lives entirely in asserting the *value* is `Ok(())`. The doc says this is load-bearing; it is, and now by measurement rather than argument.

That's this repo's recurring shape (#1817) in its inverted, *correct* form: "resolved" is one signal carrying two states, and the test refuses to conflate them. Most instances in that catalogue are guards that collapsed two states into one by accident. This is one that was deliberately built not to, and the mutant is what shows the difference.

**2. The generation genuinely failed, and the test still discriminated.** The mutant's error is a real package-input failure (`application.attention_mask` not supplied) — the fixture never completes. The engine-side test recorded the opposite conclusion, that an end-to-end version would need a package the interpreter can finish. It doesn't, and the reason is structural rather than lucky: the admission signal *precedes* the work, and the error arm's `take()` is what separates the two cases. Given the callback, admission `take()`s the sender and the error arm finds `None`. Denied it, the sender survives to the error arm and sends `Err`. So the failure of the generation is exactly what makes the mutant observable, instead of what makes the test unwritable.

That inverts the usual cost argument: this guard is cheap and non-flaky *because* it doesn't depend on the happy path, and a more "realistic" end-to-end version would be strictly worse — slower, fixture-dependent, and no more discriminating.

## Why this needed a second party at all

@Gaff flagged on #1900 that `on_admitted` had come to carry two contracts — *"the scheduler admitted this"* to the engine and *"you may begin streaming"* to the server — and warned that any resolution touching one needed a companion test pinning the other, or the two would diverge again under a green suite. This is that companion, and the asymmetry is worth stating plainly: **the engine's guards cannot see this defect, because the defect is the driver not handing over a callback at all.** An engine test asserts the engine invokes what it is *given*. Passing `None` is upstream of that. Nothing in the engine crate is capable of failing.

So the coverage boundary here is a crate boundary, and it is the kind that reads as covered from either side.

No findings. Recorded because a guard whose whole value is a mutant nobody re-runs is a guard on its way to becoming decoration — and this one earns its place.

— gaff-1

**Comment by justinchuby** [comment]
### The driver-side guard is live and non-vacuous on current `main` — re-verified, so it needn't be re-derived

This came up again as an apparently-unclaimed gap, so recording the measurement publicly rather than in a thread.

The companion test closing this issue exists and is on `main`:

```
crates/onnx-genai-server/src/driver.rs:2995
    async fn a_tensor_bound_request_is_told_it_was_admitted()
added by 2c4e60373  (#2000)
```

It lives **inside `mod admission_tests`** (`driver.rs:2359`) and calls `run_generation` directly — worth stating because that module is easy to read as covering only `route_continuous_admissions`, and a reader who stops there concludes the tensor-binding branch has neither direction under test.

Verified on `66ca26a5d`, not remembered:

```
$ cargo test -q -p onnx-genai-server --lib a_tensor_bound_request_is_told_it_was_admitted
test result: ok. 1 passed; 0 failed; 337 filtered out; finished in 0.24s
```

**It still discriminates.** Mutating `driver.rs:2276` — the `(Some(bound), session)` arm — from `Some(&mut admitted)` to `None`:

```
test result: FAILED. 0 passed; 1 failed; finished in 0.24s
panicked at driver.rs:3056:
  a tensor-bound request that the scheduler admitted must be told so,
  got Err(DriverFailure { message: "required workflow package input
  'application.attention_mask' was not supplied", kind: Internal })
```

Tree restored afterwards.

**The part that is easy to get wrong if this is rewritten:** under that mutation the oneshot still **resolves** — with `Err`. A test asserting only that the receiver completed passes under the defect. Asserting the *value* `Ok(())` is what separates the two cases, because admission precedes the work and the error arm's `take()` means a fired callback leaves `None` behind for it. That is also why the test needs no runnable package and tolerates the generation failing, which is the opposite of the conclusion the engine-side test recorded.

Nothing outstanding here.
